cpp/src/arrow/compute/exec/window_functions/bit_vector_navigator.h

```diff
@@ -0,0 +1,513 @@
+      int64_t word_end = std::min(word_count(), word_begin + words_per_block);
+
+      count += mid_counters[apply_stride_and_offset(word_end - 1)];
+      count += ARROW_POPCOUNT64(words[apply_stride_and_offset(word_end - 1)]);
+    }
+    counters[apply_stride_and_offset(block_index_end - 1)] = count;
+  }
+
+  // Position of the nth bit set (input argument zero corresponds to the first
+  // bit set).
+  //
+  int64_t Select(int64_t rank) const {
+    if (rank < 0) {
+      return BeforeFirstBit();
+    }
+    if (rank >= pop_count()) {
+      return AfterLastBit();
+    }
+
+    constexpr int64_t bits_per_block = BitVectorWithCountsBase::kBitsPerBlock;
+    constexpr int64_t bits_per_word = BitVectorWithCountsBase::kBitsPerWord;
+    constexpr int64_t words_per_block = bits_per_block / bits_per_word;
+    const int64_t* top_counters = container_->top_counts_.data();
+    const uint16_t* mid_counters = container_->mid_counts_.data();
+    const uint64_t* words = container_->bits_.data();
+
+    // Binary search in top level counters.
+    //
+    // Equivalent of std::upper_bound() - 1, but not using iterators.
+    //
+    int64_t begin = 0;
+    int64_t end = block_count();
+    while (end - begin > 1) {
+      int64_t middle = (begin + end) / 2;
+      int reject_left_half =
+          (rank >= top_counters[apply_stride_and_offset(middle)]) ? 1 : 0;
+      begin = begin + (middle - begin) * reject_left_half;
+      end = middle + (end - middle) * reject_left_half;
+    }
+
+    int64_t block_index = begin;
+    rank -= top_counters[apply_stride_and_offset(begin)];
+
+    // Continue with binary search in intermediate level counters of the
+    // selected block.
+    //
+    begin = block_index * words_per_block;
+    end = std::min(word_count(), begin + words_per_block);
+    while (end - begin > 1) {
+      int64_t middle = (begin + end) / 2;
+      int reject_left_half =
+          (rank >= mid_counters[apply_stride_and_offset(middle)]) ? 1 : 0;
+      begin = begin + (middle - begin) * reject_left_half;
+      end = middle + (end - middle) * reject_left_half;
+    }
+
+    int64_t word_index = begin;
+    rank -= mid_counters[apply_stride_and_offset(begin)];
+
+    // Continue with binary search in the selected word.
+    //
+    uint64_t word = words[apply_stride_and_offset(word_index)];
+    int pop_count_prefix = 0;
+    int bit_count_prefix = 0;
+    const uint64_t masks[6] = {0xFFFFFFFFULL, 0xFFFFULL, 0xFFULL, 0xFULL, 0x3ULL, 0x1ULL};
+    int bit_count_left_half = 32;
+    for (int i = 0; i < 6; ++i) {
+      int pop_count_left_half =
+          static_cast<int>(ARROW_POPCOUNT64((word >> bit_count_prefix) & masks[i]));
+      int reject_left_half = (rank >= pop_count_prefix + pop_count_left_half) ? 1 : 0;
+      pop_count_prefix += reject_left_half * pop_count_left_half;
+      bit_count_prefix += reject_left_half * bit_count_left_half;
+      bit_count_left_half /= 2;
+    }
```

Comment:
This can be done faster with tzcnt and pdep instructions wherever available except zen2 (pdep is slow there). Otherwise the binary search could look like above for a tighter loop.